cylc/flow/workflow_files.py

```diff
@@ -1036,18 +1004,149 @@ def get_platforms_from_db(run_dir):
+        reg = os.getcwd()
+    reg_path = Path(reg)
+    if src:
+        cylc_run_dir = get_workflow_run_dir('')
+        with suppress(ValueError):
+            reg_path = reg_path.relative_to(cylc_run_dir)
+    abs_path = (
+        reg_path if reg_path.is_absolute()
+        else Path(get_workflow_run_dir(reg_path))
+    )
+    if require_flow_file and not os.path.lexists(abs_path):
+        raise WorkflowFilesError(f"{os.strerror(errno.ENOENT)}: {abs_path}")
+    if abs_path.is_file():
+        if not src:
+            raise WorkflowFilesError(
+                f"Workflow name must refer to a directory, not a file: "
+                f"{reg_path}"
+            )
+        reg_path = reg_path.parent
+    else:
+        latest_run = infer_latest_run(abs_path)
+        if abs_path != latest_run:
+            if not is_relative_to(latest_run, abs_path):
+                reg_path = reg_path.parent
+            reg_path = reg_path / latest_run.name
+            abs_path = latest_run
+        if require_flow_file:
+            abs_path = check_flow_file(abs_path)
+    reg = str(reg_path)
+    if src:
+        return (reg, abs_path)
+    return reg
+
+
+def validate_workflow_name(name: str) -> None:
     """Check workflow name is valid and not an absolute path.
 
     Raise WorkflowFilesError if not valid.
     """
-    is_valid, message = WorkflowNameValidator.validate(flow_name)
+    is_valid, message = WorkflowNameValidator.validate(name)
     if not is_valid:
         raise WorkflowFilesError(
-            f"invalid workflow name '{flow_name}' - {message}")
-    if os.path.isabs(flow_name):
+            f"invalid workflow name '{name}' - {message}"
+        )
+    if os.path.isabs(name):
         raise WorkflowFilesError(
-            f"workflow name cannot be an absolute path: {flow_name}")
+            f"workflow name cannot be an absolute path: {name}"
+        )
+    name = os.path.normpath(name)
+    if name.startswith('.'):
+        raise WorkflowFilesError(
+            "Workflow name cannot be a path that points to the cylc-run "
+            "directory or above"
+        )
+
+
+def infer_latest_run(path: Path) -> Path:
+    """Infer the numbered run dir if the workflow has a runN symlink.
+
+    Args:
+        path: Absolute path to the workflow dir, run dir or runN dir.
+
+    Raises WorkflowFilesError if the runN symlink is not valid.
+    """
+    if path.name == WorkflowFiles.RUN_N:
+        runN_path = path
+    else:
+        runN_path = path / WorkflowFiles.RUN_N
+        if not os.path.lexists(runN_path):
+            return path
+    if not runN_path.is_symlink() or not runN_path.is_dir():
+        raise WorkflowFilesError(
+            f"runN directory at {runN_path} is a broken or invalid symlink"
+        )
+    numbered_run = os.readlink(str(runN_path))
+    if not re.match(r'run\d+$', numbered_run):
+        # Note: the link should be relative
+        # TODO: what about cylc 8.0b1 workflows where it was absolute?
```

Comment:
Should I tackle this or should we just leave it and not support the 8.0b1 absolute path targets of `runN` symlinks?

Reply:
I don't think we need to worry about 8.0b1 compat.